### PR TITLE
Fix more than 64 cpu

### PR DESCRIPTION
**Pull request by tpoisonooo**
#### 主题
尝试修复 CPU 个数大于 64 时设置亲和力失败的问题。

a. 如果是 linux/unix 流派系统，引用系统库`sched.h`的`cpu_set_t`;
b. 如果是 windows 系统，自己糙一个 `struct cpu_set_t`。装作 64 核 windows 跑 ncnn 的场景不存在...

#### 测试
跑了 benchmark 没崩
图片

### CODE REVIEW

File: src/cpu.h [hunk 1]
```diff
@@ -17,6 +17,34 @@
...
+
+#define CPU_CLR(cpu, cpusetp) \
+    ((cpusetp)->__bits[(cpu) / __NCPUBITS] &= (!(1UL << ((cpu) % __NCPUBITS))))
+
+#define CPU_ISSET(cpu, cpusetp) \
+    ((cpusetp)->__bits[(cpu) / __NCPUBITS] & (1UL << ((cpu) % __NCPUBITS)))
+
+} // namespace ncnn
+#endif
+
```
Comment (nihui) [inline src/cpu.h:47]: Linux 因为是纯C的项目，没有办法才 CPU_SET CPU_ZERO 这样
ncnn 做个 class CpuSet

Comment (tpoisonooo) [reply src/cpu.h:47]: 你是说：
```
class CpuSet {
#if android or linux
调用系统的 cpu_set_t 相关的 CPU_SET 等
#else
糙自己的 cpu_set_t
#endif
};
```
对外提供不区分平台的接口？

File: src/cpu.h [hunk 2]
```diff
@@ -16,9 +16,73 @@
...
+{
+private:
+#if defined __ANDROID__ || defined __linux__
+    cpu_set_t m_bits;
+#else
+#define CPU_SETSIZE   1024
+#define NCNN_NCPUBITS (8 * sizeof(unsigned long))
+    unsigned long m_bits[CPU_SETSIZE / NCNN_NCPUBITS];
+#endif
+
```
Comment (nihui) [inline src/cpu.h:37]: move private block after the public block

File: src/cpu.h [hunk 3]
```diff
@@ -16,9 +16,73 @@
...
+public:
+    void* data_ptr()
+    {
+        return static_cast<void*>(&m_bits);
+    }
+
+    size_t data_size()
+    {
+        return sizeof(m_bits);
+    }
```
Comment (nihui) [inline src/cpu.h:47]: These two functions are implementation-specific, and for internal use only.
Move them to protected block, and add related friend functions

File: src/cpu.h [hunk 4]
```diff
@@ -16,9 +16,73 @@
...
 
+class CpuSet
+{
+private:
+#if defined __ANDROID__ || defined __linux__
+    cpu_set_t m_bits;
+#else
+#define CPU_SETSIZE   1024
+#define NCNN_NCPUBITS (8 * sizeof(unsigned long))
+    unsigned long m_bits[CPU_SETSIZE / NCNN_NCPUBITS];
```
Comment (nihui) [inline src/cpu.h:35]: rename CPU_SETSIZE to NCNN_CPU_SETSIZE like how NCNN_NCPUBITS does

File: src/cpu.h [hunk 5]
```diff
@@ -45,10 +109,10 @@ int get_cpu_powersave();
 int set_cpu_powersave(int powersave);
 
 // convenient wrapper
-size_t get_cpu_thread_affinity_mask(int powersave);
+CpuSet get_cpu_thread_affinity_mask(int powersave);
 
 // set explicit thread affinity
-int set_cpu_thread_affinity(size_t thread_affinity_mask);
+int set_cpu_thread_affinity(CpuSet thread_affinity_mask);
```
Comment (nihui) [inline src/cpu.h:115]: ```const CpuSet&```

File: src/cpu.h [hunk 6]
```diff
@@ -45,10 +109,10 @@ int get_cpu_powersave();
 int set_cpu_powersave(int powersave);
 
 // convenient wrapper
-size_t get_cpu_thread_affinity_mask(int powersave);
+CpuSet get_cpu_thread_affinity_mask(int powersave);
```
Comment (nihui) [inline src/cpu.h:112]: The global affinity masks are always defined, return ```const CpuSet&```

File: src/cpu.h (hunk 2 above)
Comment (nihui) [inline src/cpu.h:37]: The constructor is missing

**Comment by tpoisonooo** [comment]
done.

**Comment by tpoisonooo** [comment]
@nihui  重新触发一下  CI ？ 我就merge 一下 code format 就不行了...

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=1973) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=1973) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: tpoisonooo<br/>:x: restyled-commits<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/ncnn?pullRequest=1973) it.</sub>

**Comment by tpoisonooo** [comment]
close 大法好！

**Comment by tpoisonooo** [comment]
@nihui  CI 里面 
```
The command "eval git clone --depth=50 --branch=restyled/pull-1973 https://github.com/Tencent/ncnn.git Tencent/ncnn " failed. Retrying, 2 of 3.

求看一下。

File: src/cpu.cpp [hunk 7]
```diff
@@ -440,7 +411,7 @@ int set_cpu_powersave(int powersave)
         return -1;
     }
 
-    size_t thread_affinity_mask = get_cpu_thread_affinity_mask(powersave);
+    CpuSet thread_affinity_mask = get_cpu_thread_affinity_mask(powersave);
```
Comment (nihui) [inline src/cpu.cpp:414]: const CpuSet&

File: src/cpu.h [hunk 8]
```diff
@@ -16,9 +16,109 @@
...
+#define CPU_ZERO(cpusetp) \
+    memset((cpusetp), 0, sizeof(cpu_set_t))
+
+#define CPU_CLR(cpu, cpusetp) \
+    ((cpusetp)->__bits[(cpu) / __NCPUBITS] &= (!(1UL << ((cpu) % __NCPUBITS))))
+
+#define CPU_ISSET(cpu, cpusetp) \
+    ((cpusetp)->__bits[(cpu) / __NCPUBITS] & (1UL << ((cpu) % __NCPUBITS)))
+
+} // namespace ncnn
```
Comment (nihui) [inline src/cpu.h:50]: 这些定义似乎并没有哪里用到了

File: src/cpu.h [hunk 9]
```diff
@@ -16,9 +16,109 @@
...
+#define NCNN_CPU_SETSIZE 1024
+#define NCNN_NCPUBITS    (8 * sizeof(unsigned long))
+class CpuSet
+{
+public:
+    CpuSet()
+    {
+#if defined __ANDROID__ || defined __linux__
+        CPU_ZERO(&m_bits);
+#endif
```
Comment (nihui) [inline src/cpu.h:64]: 这里直接调用 zero()

File: src/cpu.h [hunk 10]
```diff
@@ -16,9 +16,109 @@
...
+    size_t data_size() const
+    {
+        return sizeof(m_bits);
+    }
+
+private:
+#if defined __ANDROID__ || defined __linux__
+    cpu_set_t m_bits;
+#else
+    unsigned long m_bits[NCNN_CPU_SETSIZE / NCNN_NCPUBITS] = {0};
```
Comment (nihui) [inline src/cpu.h:119]: 如果 android / linux 以外的平台都是没有用的，就没有必要用 m_bits 模拟这个了

**Review by nihui — CHANGES_REQUESTED** [review]

File: src/cpu.h [hunk 11]
```diff
@@ -16,9 +16,65 @@
 #define NCNN_CPU_H
 
 #include <stddef.h>
+#if defined __ANDROID__ || defined __linux__
+#include <sched.h>
+#endif
+
+#if defined __ANDROID__ || defined __linux__
+#include <sched.h>
+#endif
```
Comment (nihui) [inline src/cpu.h:25]: duplicated includes

**Review by nihui — CHANGES_REQUESTED** [review]

**Pull request closed without merging**